vncviewer/touch.cxx

```diff
@@ -0,0 +1,324 @@
+/* Copyright 2019 Pierre Ossman <ossman@cendio.se> for Cendio AB
+ * Copyright 2019 Aaron Sowry for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <assert.h>
+
+#include <map>
+
+#if defined(WIN32)
+#include <windows.h>
+#include <commctrl.h>
+#elif !defined(WIN32) && !defined(__APPLE__)
+#include <X11/extensions/XInput2.h>
+#include <X11/extensions/XI2.h>
+#endif
+
+#include <FL/Fl.H>
+#include <FL/x.H>
+
+#include <rfb/LogWriter.h>
+
+#include "i18n.h"
+#include "vncviewer.h"
+#include "BaseTouchHandler.h"
+#if defined(WIN32)
+#include "win32TouchHandler.h"
+#elif defined(__APPLE__)
+#else
+#include "XInputTouchHandler.h"
+#endif
+
+#include "touch.h"
+
+static rfb::LogWriter vlog("Touch");
+
+#if !defined(WIN32) && !defined(__APPLE__)
+static int xi_major;
+#endif
+
+typedef std::map<Window, class BaseTouchHandler*> HandlerMap;
+static HandlerMap handlers;
+
+#if defined(WIN32)
+LRESULT CALLBACK win32_gestureProc(HWND hWnd, UINT uMsg, WPARAM wParam,
+                                   LPARAM lParam, UINT_PTR uIdSubclass,
+                                   DWORD_PTR dwRefData)
+{
+  GESTUREINFO gi;
+
+  DWORD blockOneFingerPan = GC_PAN_WITH_SINGLE_FINGER_VERTICALLY |
+                            GC_PAN_WITH_SINGLE_FINGER_HORIZONTALLY;
+  GESTURECONFIG gc[] = {{GID_ZOOM, GC_ZOOM, 0},
+                        {GID_PAN, GC_PAN, blockOneFingerPan},
+                        {GID_TWOFINGERTAP, GC_TWOFINGERTAP, 0}};
+
+  switch(uMsg) {
+
+  case WM_GESTURE:
+    ZeroMemory(&gi, sizeof(GESTUREINFO));
+    gi.cbSize = sizeof(GESTUREINFO);
+
+    if (!GetGestureInfo((HGESTUREINFO)lParam, &gi)) {
+      vlog.error(_("Failed to get gesture information (error 0x%i)"),
```

Comment:
`0x%i`?!

Reply:
🤦‍♂ fixed in 016d8328842bcec8c5b6d4fe33f981ba5700f545